src/contrib/mlir/core/ngraph_dialect/ops.cpp

```diff
@@ -328,6 +328,28 @@ mlir::IntegerAttr getBufferId(mlir::Operation* op)
     return op->getAttrOfType<mlir::IntegerAttr>(getBufferIdAttrName());
 }
 
+static std::string getBufferOffsetAttrName()
+{
+    return "ng.buffer_offset";
+}
+
+void setBufferOffset(mlir::Operation* op, mlir::IntegerAttr attr)
+{
+    op->setAttr(getBufferOffsetAttrName(), attr);
+}
+
+mlir::IntegerAttr setBufferOffset(mlir::Operation* op, unsigned val)
+{
+    auto attr = mlir::IntegerAttr::get(IntegerType::get(32, op->getContext()), val);
+    setBufferOffset(op, attr);
+    return attr;
+}
```

Comment:
Have you considered storing this information in a separate analysis class? I think that adding analysis information to the main IR doesn't scale and it's more difficult to update/invalidate/retrieve the information since we have to traverse the IR for that. An analysis that contains this information in a map<Operation *, BufferInfo> could work pretty well and it would make update/invalidate/retrieve very simple without modifying the main IR. Some examples of this in LLVM, like SCEV.

Reply:
I see your point. But it would be difficult to invalidate the map due to IR modifications.  Generally, attaching info to the IR is better IMO, because if a value is modified/deleted we can detect the invalid state, and bail out.  
For example, consider
```
X = ...
V = Concat X, Y, Z

The side map would say:
Concat --> BufferInfo{ Operand0-> buffer0, offset: 16, Operand1--> ...}
```
If the value of X is modified or replaced, for example the operation is decomposed, the information in the map will be invalid, and we wouldn't be able to tell. Yielding to incorrect in-place concat. 

OTH if the info is attached to X's def directly, the attribute will just be removed, and when we try to do the concat in-place, we will detect the missing data, and just bail out.

Reply:
The example that you describe is how analysis works in LLVM in general. When a pass modifies the IR, it invalidates any prior analysis unless the pass knows how to preserve/update it. You can see how vectorizer updates loop info, dominator tree and scalar evolution (all storing analysis information separately) after IR modification here: https://llvm.org/doxygen/LoopVectorize_8cpp_source.html#l04338. Scalar evolution, for example, uses a map to store SCEV for each Instruction (http://llvm.org/doxygen/ScalarEvolution_8h_source.html#l01181), which is pretty similar to what we would need here. Among other things, storing analysis information to the IR is problematic because we would have to traverse the whole IR for invalidation, cloning the IR would clone all the analysis information for that IR (now invalid for the clone?), and analysis data structures may become corrupted after a change in the IR if they point to info stored in the IR. It's kind of separation of concerns principle.